viewing_party/party.py

```diff
@@ -1,23 +1,151 @@
 # ------------- WAVE 1 --------------------
+from hashlib import new
+# from turtle import title
+from tests.test_constants import MOVIE_TITLE_1, clean_wave_3_data
+from collections import Counter
+
 
 def create_movie(title, genre, rating):
-    pass
 
+    if title == None or genre == None or rating == None:
+        return None
+    else:
+        movie = {
+                        "title": title,
+                        "genre": genre,
+                        "rating": rating,
+                        }
+        return movie
+    
+
+def add_to_watched(user_data, movie):
+   
+    user_data["watched"].append(movie)
+    return user_data
+
+
+def add_to_watchlist(user_data, movie):
+    
+    user_data["watchlist"].append(movie)
+    return user_data
+
+
+def watch_movie(user_data, title):
+
+    to_watch_list = user_data["watchlist"]
+    already_watched = user_data["watched"]
+
+    for i in range(len(to_watch_list)):
+        if to_watch_list[i]["title"] == title:
+            already_watched.append(to_watch_list[i])
+            to_watch_list.remove(to_watch_list[i])
+    return user_data
+    
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+   
+    already_watched = user_data["watched"]
+    ratings = 0
+    for i in range(len(already_watched)):
+        ratings += already_watched[i]["rating"]
+        avr_rating = (ratings/(len(already_watched)))
+    if already_watched == []:
+        avr_rating = 0
+        
+    return avr_rating
+
+def get_most_watched_genre(user_data):
 
+     already_watched = user_data["watched"]
+     all_genres = []
+
+     for i in range(len(already_watched)):
+        all_genres.append(already_watched[i]["genre"])
+        top_genre = [word for word, word_count in Counter(all_genres).most_common(1)]
+        top_genre = top_genre[0]
+        return top_genre
+     if already_watched == []:
+        top_genre = None
+     return top_genre
+     
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    
+    friends_not_watched = []
+    friends_watched = []
+    user_watched = user_data["watched"]
+
+    for movie in user_watched:
+        friends_not_watched.append(movie)
+
+    for friend_list in user_data["friends"]:
+        for friend in friend_list["watched"]:
+            friends_watched.append(friend)
+    for i in friends_watched:
+        if i in friends_not_watched:
+            friends_not_watched.remove(i)

```

Comment:
These loop variable names are a little confusing. Does `friend_list` really represent an entire list? Or an item within a list? And does `friend` really represent every item in the `watched` list? 

They say that there 2 hard problems in computer science: cache invalidation, naming things, and off-by-one errors! Using the debugger or even doing a simple print statement within a for loop to see what the loop variable represents can really help when deciding variable names.

Reply:
We can also combine these two for loops and check if the `friend` is already inside of the `friend_not_watched` list. 

```python
for friend_list in user_data["friends"]:
   for friend in friend_list["watched"]:
        if i in friends_not_watched:     
            friends_watched.append(friend)
```